Dev mode

odysseus0 opened the pull request.

odysseus0 committed:
Refactor main.rs and add configuration management

- Add figment dependency for configuration management
- Create new config module to handle configuration loading
- Refactor main.rs to use new Config struct and improve code organization
- Extract helper functions for better modularity
- Update environment variables in teleport.env
- Remove hardcoded paths and use Config struct for file paths

odysseus0 committed:
Update .gitignore and add environment file examples

- Modified .gitignore to exclude all *.env files except *.env.example
- Removed specific .env files from version control
- Added example environment files for staging and production

odysseus0 committed:
Update Dockerfile and docker-compose with comments

odysseus0 committed:
Enhance README with quickstart guide

odysseus0 committed:
docs: Add ARCHITECTURE.md with project overview

Create a detailed architecture document that includes:
- High-level project overview
- Key components and project structure
- Detailed explanations of core processes:
  - NFT minting and redemption
  - TEE attestation
  - Authorization flow
  - Content filtering
- Security considerations and future improvements

This document serves as a central reference for understanding
the Teleport project's design and implementation.

odysseus0 committed:
docs: Add comprehensive DEVELOPMENT.md guide

Create a detailed development guide to assist contributors with:
- Environment setup and prerequisites
- Docker configuration and usage
- Building and running the project
- Debugging instructions, including SGX considerations
- Code style guidelines
- Common issues and troubleshooting
- Useful commands for development
- Links to relevant documentation and resources

This guide aims to streamline the onboarding process for new developers
and serve as a reference for existing team members.

odysseus0 committed:
Added the development production flag and adjusted the main function accordingly

odysseus0 committed:
Configure Rust to use group imports and run formatting on all files

- Updated the Rust configuration to enable grouping of imports.
- Ran `cargo fmt` to format all Rust files in the project.

odysseus0 committed:
feat: add dev mode command to Makefile

odysseus0 committed:
Refactor variable names for clarity and consistency

- Updated variable names from `pkey` to `private_key` in the production setup to enhance readability and maintainability.
- Modified the `cookietest` function to use underscore-prefixed variables for unused parameters, following Rust conventions.
- Changed logging in the development server to use `log::info` instead of `println!` for better logging practices.

These changes improve code clarity and adhere to best practices in variable naming and logging.

odysseus0 committed:
Enhance Production Configuration and Refactor Main Logic

- Updated `.vscode/settings.json` to include `rust-analyzer.cargo.features` for production.
- Modified `Cargo.toml` to set the default feature to `production`.
- Refactored `src/main.rs`:
  - Cleaned up imports by removing unused ones and organizing them.
  - Changed the `main` function signature to return a `Result` for better error handling.
  - Updated server setup logic to use a new `server_setup` module for both production and development environments.
  - Ensured proper handling of private keys and CSR creation.
  - Removed redundant server setup functions and integrated their logic into the new module.

odysseus0 committed:
Refactor SGX Attestation Handling

- Updated the call to `handle_sgx_attestation` to use the new path `sgx_attest::handle_sgx_attestation`.
- Moved the `handle_sgx_attestation` function from `src/main.rs` to `src/sgx_attest.rs` for better organization.
- Ensured the function is now asynchronous and uses `tokio::fs::write` for file operations.
- Added necessary imports for `PKey` and `X509Req` in `src/sgx_attest.rs`.

This change improves code modularity and maintains the functionality of SGX attestation handling.

odysseus0 committed:
Refactor main.rs and introduce router module

- Move app creation logic to a new router module
- Update imports and remove unused ones
- Reorganize module declarations for better readability

odysseus0 committed:
Refactor database module and improve shutdown handling

- Reorganize db module structure:
  - Move data models to new 'models.rs' file
  - Create 'utils.rs' for database utility functions
- Simplify main.rs:
  - Move db loading/saving logic to utils
  - Improve shutdown process with dedicated function
- Remove unused imports and clean up code

odysseus0 committed:
refactor: reformat code

odysseus0 committed:
Refactor certificate handling and SGX attestation

- Move certificate-related functions from main.rs to cert.rs
- Update main.rs to use the new certificate functions from cert.rs
- Remove redundant imports in main.rs
- Improve code organization and modularity

odysseus0 committed:
Configure VSCode settings for Rust development

- Enable format on save
- Adjust Rust Analyzer cargo features formatting

odysseus0 committed:
feat: improve private.env docs and add teleport.env docs

odysseus0 committed:
docs(db): improve documentation for TeleportDB trait

- Add detailed documentation comments for all methods in the TeleportDB trait
- Clarify the purpose, arguments, and return values for each method
- Include TODO comments for future improvements (e.g., stronger typing for addresses and IDs)
- Add explanatory comments for data structures in InMemoryDB

odysseus0 committed:
refactor: improve configuration and server setup

- Add Environment enum to Config for development/production distinction
- Refactor server setup to use separate functions for dev and prod
- Remove feature flags, use Environment enum instead
- Update TwitterBuilder to take string slices
- Add serde derive feature
- Minor code cleanup and formatting improvements

odysseus0 committed:
feat: add .cursorrules

odysseus0 committed:
feat(config): Update default configuration and environment variables

- Added new configuration options in `config/default.toml` for TEE_URL, APP_URL, environment, and DB_PATH to set good default values.
- Updated documentation in `DEVELOPMENT.md` to reflect changes in configuration management.
- Improved clarity on the purpose of environment variables and their usage.
- Ensured sensitive information is properly documented and emphasized the importance of configuration files.

odysseus0 committed:
fix: run markdown lint to correct formatting issues in README.md

odysseus0 committed:
fix: resolve Clippy linting issues

- Updated the Tweet initialization to use `to_string()` for clarity.
- Changed NFT struct references to use the correct naming convention (Nft instead of NFT).
- Adjusted function signatures in the TeleportDB trait and InMemoryDB implementation to reflect the updated Nft type.
- Fixed parameter passing in the setup_production_server function for consistency.

These changes improve code readability and adhere to Rust's naming conventions.

amiller commented on .gitignore:
```diff
@@ -1,13 +1,14 @@
\ No newline at end of file
+tokens.json
+
+*.env
```
teleport.env should not be ignored since it contains public data to bake into the enclave

amiller commented on Cargo.toml:
```diff
@@ -24,15 +29,15 @@ serde_qs = "0.13.0"
+tower-http = { version = "0.5.2", features = ["cors"] }
 http = "1.1.0"
-axum-server = {version = "0.6.0", features = ["tls-rustls"]}
+axum-server = { version = "0.6.0", features = ["tls-rustls"] }
```
would prefer to review a separate PR for cosmetic changes just to keep the scope of changed files smaller

amiller commented on docker-compose.yml:
```diff
@@ -7,15 +7,14 @@ services:
-      - /dev/sgx_enclave
-      - /dev/sgx_provision
+      - /dev/sgx_enclave  # SGX enclave device
+      - /dev/sgx_provision  # SGX provisioning device
```
these aren't adding info

amiller commented on docs/ARCHITECTURE.md:
```diff
@@ -0,0 +1,90 @@
+# Teleport Architecture
```
I don't like using Gramine debug mode. Remote attestation etc is unavailable. I would rather just not use Gramine when running outside enclave

amiller reviewed (changes requested):
Spent a little bit of time on it but theres lots of documentation changes and I'm not sure what substantive ones, I'd want to review them separately basically

odysseus0 committed:
refactor(config): implement structured configuration management

- Replace figment with config crate for more flexible configuration handling
- Introduce environment-specific config files (development, staging, production)
- Remove teleport.env and integrate its variables into the new config structure
- Update Config struct to reflect new configuration layout
- Modify main.rs to use the new configuration setup
- Add strum and strum_macros for better enum handling
- Update documentation in DEVELOPMENT.md to reflect new configuration approach

odysseus0 committed:
refactor(config): remove now unnecessary variables

odysseus0 committed:
refactor(config): improve environment handling and logging

- Rename `environment` to `env` in Config struct for consistency
- Load environment variables from "private.env" file
- Change environment variable name from RUN_MODE to APP_ENV
- Enhance configuration logging with pretty-print format
- Update main.rs to use the new `env` field name

odysseus0 committed:
feat: Update development configuration and documentation

- Added ENV variable to development.toml to specify the environment as "development".
- Updated APP_URL in development.toml to include the protocol (http).
- Enhanced DEVELOPMENT.md to include instructions for running the server in development mode without TEE, Gramine, TLS, or Docker.
- Recommended enabling logging during development for better visibility.

odysseus0 committed:
fix: update URLs in configuration files

- Removed the "http://" and "https://" prefixes from APP_URL and TEE_URL in development, production, and staging configuration files.
- This change ensures consistency in URL formatting across environments.

odysseus0 committed:
feat(auth): add logging for OAuth token request

Added logging statements to the `request_oauth_token` function to log the request for the OAuth token and the callback URL being used. This improves traceability and debugging capabilities during the OAuth flow.

odysseus0 committed:
refactor: improve configuration management and reduce global state

- Replace `config` crate with `figment` for more flexible configuration
- Restructure configuration into nested structures (AppConfig, Secrets)
- Remove environment variables used as global state in functions
- Remove instances of using env vars inside module functions as global variables
- Update main.rs to use new configuration structure
- Adjust related functions in actions/nft.rs, endpoints.rs, and oai.rs

odysseus0 committed:
feat(docs): update DEVELOPMENT.md for improved configuration clarity

- Enhanced the description of the configuration system to clarify the layered approach.
- Updated the order and details of configuration sources, including the introduction of `config.toml` and its structure.
- Added instructions for setting the `APP_ENV` environment variable and its impact on configuration.
- Emphasized the importance of the `private.env` file for sensitive information and provided guidance on its creation.
- Included references to relevant files for better navigation and understanding.

odysseus0 committed:
refactor: replace openai-rust with async-openai

- Replace openai-rust crate with async-openai for improved API interaction
- Refactor OpenAI client implementation for better error handling and type safety
- Update NFT event subscription to use new OpenAI client
- Adjust content moderation logic to use structured responses
- Update dependencies in Cargo.toml
- Minor code cleanup and type improvements

odysseus0 committed:
feat: add coding standard as Cursor Rules

odysseus0 committed:
refactor(nft): move NFT-related code to separate module

- Create new `nft_events` module
- Move NFT event handling and subscription logic from `actions/nft.rs` to `nft_events/services.rs`
- Update imports and function calls in `main.rs` and `endpoints.rs`
- Remove `actions/nft.rs` file
- Minor adjustments to `SharedState` and `TwitterBuilder` usage

odysseus0 committed:
refactor(nft): reorganize NFT-related modules and update imports

- Rename 'nft_events' directory to 'nft'
- Move handlers, services, and subscription logic into separate files
- Update import paths in main.rs and endpoints.rs
- Improve code organization and modularity for NFT-related functionality

odysseus0 committed:
feat(config): add scheme configuration for URLs

- Add 'scheme' field to AppConfig struct in config.rs
- Implement Scheme enum with Http and Https variants
- Update config.toml with scheme settings for different environments
- Modify get_callback_url function to use the configured scheme
- Enhance error handling and logging in callback endpoint
- Add debug derive for query structs in endpoints.rs

This commit improves URL handling by allowing configurable HTTP/HTTPS schemes
and enhances error handling and logging in the callback process.

odysseus0 committed:
feat(deps): upgrade alloy to 0.4.1 and replace cuid with cuid2

- Update alloy dependency from 0.3.1 to 0.4.1
- Replace cuid with cuid2 for ID generation
- Add TODO comment about potentially switching to UUID in the future
- Update Cargo.lock with new dependency versions

amiller commented on src/twitter/mod.rs:
```diff
@@ -9,10 +11,11 @@ pub fn get_callback_url(
     callback_base_url: String,
     address: String,
     frontend_nonce: String,
+    scheme: &Scheme,
```
this suggests there is some generality in the code for https/http. But given that http is only approprirate for development, it would be better to make this more conspicuous

amiller commented on src/server_setup.rs:
```diff
@@ -0,0 +1,46 @@
+    Ok(())
+}
+
+pub async fn setup_development_server(app: Router) -> Result<(), Box<dyn std::error::Error>> {
```
maybe it would be better to use https exclusively and inlcude a self-signed certificate you can use when developing and accessing localhost
